src/conf_mode/nat_cgnat.py

```diff
@@ -0,0 +1,287 @@
+            if 'range' not in pool_config:
+                raise ConfigError(
+                    f'Range for "{pool} pool {pool_name}" must be defined!'
+                )
+
+    for rule, rule_config in config['rule'].items():
+        if 'source' not in rule_config:
+            raise ConfigError(f'Rule "{rule}" source pool must be defined!')
+        if 'pool' not in rule_config['source']:
+            raise ConfigError(f'Rule "{rule}" source pool must be defined!')
+
+        if 'translation' not in rule_config:
+            raise ConfigError(f'Rule "{rule}" translation pool must be defined!')
+
+
+def generate(config):
+    if not config:
+        return None
+    # first external pool as we allow only one as PoC
+    ext_pool_name = jmespath.search("rule.*.translation | [0]", config).get('pool')
+    int_pool_name = jmespath.search("rule.*.source | [0]", config).get('pool')
+    ext_query = f"pool.external.{ext_pool_name}.range | keys(@)"
+    int_query = f"pool.internal.{int_pool_name}.range"
+    external_ranges = jmespath.search(ext_query, config)
+    internal_ranges = [jmespath.search(int_query, config)]
+
+    external_list_count = []
+    external_list_hosts = []
+    internal_list_count = []
+    internal_list_hosts = []
+    for ext_range in external_ranges:
+        # External hosts count
+        e_count = IPOperations(ext_range).get_ips_count()
+        external_list_count.append(e_count)
+        # External hosts list
+        e_hosts = IPOperations(ext_range).convert_prefix_to_list_ips()
+        external_list_hosts.extend(e_hosts)
+    for int_range in internal_ranges:
+        # Internal hosts count
+        i_count = IPOperations(int_range).get_ips_count()
+        internal_list_count.append(i_count)
+        # Internal hosts list
+        i_hosts = IPOperations(int_range).convert_prefix_to_list_ips()
+        internal_list_hosts.extend(i_hosts)
+
+    external_host_count = sum(external_list_count)
+    internal_host_count = sum(internal_list_count)
+    ports_per_user = int(
+        jmespath.search(f'pool.external.{ext_pool_name}.per_user_limit.port', config)
+    )
+    external_port_range: str = jmespath.search(
+        f'pool.external.{ext_pool_name}.external_port_range', config
+    )
+
+    proto_maps, other_maps = generate_port_rules(
+        external_list_hosts, internal_list_hosts, ports_per_user, external_port_range
+    )
+
+    config['proto_map_elements'] = ', '.join(proto_maps)
+    config['other_map_elements'] = ', '.join(other_maps)
+
+    render(nftables_cgnat_config, 'firewall/nftables-cgnat.j2', config)
+
+    # dry-run newly generated configuration
+    tmp = run(f'nft --check --file {nftables_cgnat_config}')
+    if tmp > 0:
+        raise ConfigError('Configuration file errors encountered!')
+
+
+def apply(config):
+    if not config:
+        # Cleanup cgnat
+        cmd('nft delete table ip cgnat')
+        os.unlink(nftables_cgnat_config)
```

Comment:
I'd add a condition to check if the file exists. What when a user deletes it for fun? An exception would emerge.

Reply:
Done!